parser-typechecker/src/U/Codebase/Branch/Diff.hs

```diff
@@ -113,34 +117,37 @@ instance Semigroup NameBasedDiff where
 
 -- | Diff two Branches, returning a tree containing all of the changes
-diffBranches :: forall m. (Monad m) => Branch m -> Branch m -> m TreeDiff
+diffBranches :: forall m. (Monad m) => Branch m -> Branch m -> m (TreeDiff m)
 diffBranches from to = do
   let termDiffs = diffMap (Branch.terms from) (Branch.terms to)
   let typeDiffs = diffMap (Branch.types from) (Branch.types to)
   let defDiff = DefinitionDiffs {termDiffs, typeDiffs}
-  childDiff <- do
-    Align.align (children from) (children to)
-      & wither \case
-        This ca -> do
-          -- TODO: For the names index we really don't need to know which exact
-          -- names were removed, we just need to delete from the index using a
-          -- prefix query, this would be faster than crawling to get all the deletes.
-          removedChildBranch <- Causal.value ca
-          Just . unTreeDiff <$> diffBranches removedChildBranch Branch.empty
-        That ca -> do
-          newChildBranch <- Causal.value ca
-          Just . unTreeDiff <$> diffBranches Branch.empty newChildBranch
-        These fromC toC
-          | Causal.valueHash fromC == Causal.valueHash toC -> do
-              -- This child didn't change.
-              pure Nothing
-          | otherwise -> do
-              fromChildBranch <- Causal.value fromC
-              toChildBranch <- Causal.value toC
-              diffBranches fromChildBranch toChildBranch >>= \case
-                Lens.Empty -> pure Nothing
-                TreeDiff cfr -> pure . Just $ cfr
-  pure $ TreeDiff (defDiff :< childDiff)
+  let childDiff = do
+        Align.align (children from) (children to)
+          & wither \case
+            This ca -> do
+              -- TODO: For the names index we really don't need to know which exact
+              -- names were removed, we just need to delete from the index using a
+              -- prefix query, this would be faster than crawling to get all the deletes.
+              removedChildBranch <- Causal.value ca
+              Just . unTreeDiff <$> diffBranches removedChildBranch Branch.empty
+            That ca -> do
+              newChildBranch <- Causal.value ca
+              Just . unTreeDiff <$> diffBranches Branch.empty newChildBranch
+            These fromC toC
+              | Causal.valueHash fromC == Causal.valueHash toC -> do
+                  -- This child didn't change.
+                  pure Nothing
+              | otherwise -> do
+                  fromChildBranch <- Causal.value fromC
+                  toChildBranch <- Causal.value toC
+                  diffBranches fromChildBranch toChildBranch >>= \case
+                    TreeDiff (defDiffs :< Compose mchildren) -> do
+                      children <- mchildren
+                      if (isEmptyDefDiffs defDiffs && null children)
+                        then pure Nothing
+                        else pure . Just $ (defDiffs :< Compose (pure children))
+  pure $ TreeDiff (defDiff :< Compose childDiff)
```

Comment:
It looks like `diffBranches` should return a `TreeDiff m` instead of `m (TreeDiff m)`.

Reply:
Great catch 👍🏼